File: Sistemas-Operacionais-de Tempo real/Sistemas-Operacionais-Embarcados-main/process_states.c
/*
 * process_states.c
 * Demonstra processos nos estados R, S, Z e T
 *
 * Compilação:
 *   gcc process_states.c -o process_states
 *
 * Execução:
 *   ./process_states
 * Em outro terminal:
 *   ps aux | grep process_states
 *   Use 'ps -o pid,stat,comm -p %d,%d,%d,%d' para ver os estados.
 *   top -H -p <PID>
 * 
 * Explicação: 
 * 
 * RUNNING (R): filho 1 faz loop infinito → sempre usando CPU ou pronto para usar.
 * SLEEPING (S): filho 2 chama sleep(10) repetidamente → entra em estado S.
 * ZOMBIE (Z): filho 3 termina, mas o pai não chama wait() → permanece como zumbi.
 * STOPPED (T): filho 4 é parado com SIGSTOP.
 * 
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

int main() {
    pid_t pidR, pidS, pidZ, pidT;

    printf("PID do programa principal: %d\n", getpid());

    // -----------------------------
    // 1. Processo em RUNNING (R)
    // -----------------------------
    pidR = fork();
    if (pidR == 0) {
        // Filho 1: laço infinito (sempre R ou pronto para rodar)
        while (1) {
            // pequeno trabalho para ocupar CPU
        }
        exit(0);
    }

    // -----------------------------
    // 2. Processo em SLEEPING (S)
    // -----------------------------
    pidS = fork();
    if (pidS == 0) {
        // Filho 2: dorme continuamente
        while (1) {
            sleep(10); // estado "S" (esperando)
        }
        exit(0);
    }

    // -----------------------------
    // 3. Processo em ZOMBIE (Z)
    // -----------------------------
    pidZ = fork();
    if (pidZ == 0) {
        printf("Filho Z (vai virar zumbi) PID=%d\n", getpid());
        exit(0); // termina imediatamente
    } else {
        sleep(1); // pai NÃO chama wait() → filho fica em estado Z
    }

    // -----------------------------
    // 4. Processo em STOPPED (T)
    // -----------------------------
    pidT = fork();
    if (pidT == 0) {
        while (1) {
            pause(); // espera por sinais
        }
        exit(0);
    } else {
        sleep(1); // garante que filho está em execução
        kill(pidT, SIGSTOP); // envia SIGSTOP → processo fica em T
    }

    // -----------------------------
    // Mantém o pai vivo para observar no ps/top
    // -----------------------------
    printf("Processos criados: R=%d  S=%d  Z=%d  T=%d\n",
           pidR, pidS, pidZ, pidT);

    printf("Use 'ps -o pid,stat,comm -p %d,%d,%d,%d' para ver os estados.\n",
           pidR, pidS, pidZ, pidT);

    while (1) {
        sleep(30);
    }

    return 0;
}

File: Sistemas-Operacionais-de Tempo real/Sistemas-Operacionais-Embarcados-main/process_states.py
"""
process_states.py
Demonstra processos em diferentes estados (R, S, Z, T) no Linux.

Execução:
    python3 process_states.py

Em outro terminal, observe:
    ps -o pid,stat,comm -p <pid1>,<pid2>,<pid3>,<pid4>
ou
    top -H -p <PID_DO_PAI>
"""

import os
import time
import signal

print(f"PID do programa principal: {os.getpid()}")

# --------------------------------
# 1. Processo em RUNNING (R)
# --------------------------------
pidR = os.fork()
if pidR == 0:  # filho R
    while True:
        pass  # laço infinito -> sempre R (rodando ou pronto para rodar)

# --------------------------------
# 2. Processo em SLEEPING (S)
# --------------------------------
pidS = os.fork()
if pidS == 0:  # filho S
    while True:
        time.sleep(10)  # bloqueia em sleep -> estado "S" (esperando)

# --------------------------------
# 3. Processo em ZOMBIE (Z)
# --------------------------------
pidZ = os.fork()
if pidZ == 0:  # filho Z
    print(f"Filho Z (vai virar zumbi) PID={os.getpid()}")
    os._exit(0)  # termina imediatamente
else:
    time.sleep(1)  # pai NÃO chama wait() -> filho fica como zumbi

# --------------------------------
# 4. Processo em STOPPED (T)
# --------------------------------
pidT = os.fork()
if pidT == 0:  # filho T
    while True:
        signal.pause()  # espera sinal (fica bloqueado)
else:
    time.sleep(1)  # espera o filho iniciar
    os.kill(pidT, signal.SIGSTOP)  # envia SIGSTOP -> estado T (parado)

# --------------------------------
# Mantém o pai vivo para observação
# --------------------------------
print(f"Processos criados: R={pidR}  S={pidS}  Z={pidZ}  T={pidT}")
print("Use 'ps -o pid,stat,comm -p {0},{1},{2},{3}' para ver os estados.".format(
    pidR, pidS, pidZ, pidT))

while True:
    time.sleep(30)  # mantém processo pai vivo
